src/Microsoft.OData.Client/BaseAsyncResult.cs

```diff
@@ -26,6 +27,11 @@ internal abstract class BaseAsyncResult : IAsyncResult
         /// <summary>wrapped request</summary>
         protected PerRequest perRequest;
 
+        /// <summary>
+        /// Cache used to store temporary metadata used for materialization of OData items.
+        /// </summary>
+        protected MaterializerCache materializerCache = new MaterializerCache();

```

Comment:
I think it's a good idea, but since this is an internal class, I think it would be better to instantiate the cache until there's a need to make it configurable. This way, we think of it as an implementation detail of the `BaseAsyncResult` that the caller does not need to know about. If we pass it on the constructor, the caller will have to take responsibility for passing down the cache, and we might make the argument that it should be configurable at that level, all the way up to the DataServiceContext. But the buck has to stop somewhere since this isn't a user input. I think `BaseAsyncResult` is good place because it's what glues the request together, and this cache is scoped to a request.